doc/design.md

```diff
@@ -0,0 +1,1268 @@
+- Actually `convert` to convert units is pretty specific as it needs to connect instrument information to data.
+  Is having it as part of this lowest-level library the right choice?
+
+There are some cases which are likely to lead to additions to the dataset library:
+- Initially we will not have covered all required cases with tags for `Dim` and variable tags for `Coord` and `Data`.
+  In particular, we need a new `Coord` for every unit that we support.
+- As a consequence the unit handling will also need to be extended.
+
+
+### <a name="design-details-api-stability"></a>API stability
+
+The C++ version as well as the Python exports are meant as a low level building block for other parts of the project, in addition to direct use in scripts by users.
+Therefore we need to ensure that the API is reasonably stable.
+At the same time, we will struggle with things that do not work out early on as imagined right now and will need to change the API.
+- The API is considered unstable and can change without notice until the library version has reached v1.0.
+- With v1.0 onwards the Python API should only change with major releases and a proper deprecation cycle.
+- The C++ API should ideally follow the same rules as the Python API.
+  However, since most or all C++ code using the API should be under our control, the API may be updated in a less restrictive way if there is a good reason.
+
+The implication is that this library would probably require versioning independent of Mantid itself.
+
+
+### <a name="design-details-naming"></a>Naming
+
+#### Components of the library are part of namespace `dataset`
+
+##### Rationale
+
+Mantid currently uses upper-case namespace names.
+However, this can lead to artificial and unintuitive namespace or class names.
+Instead, namespaces like `dataset` should be lower-case, as opposed to types which are upper-case.
+We then have, e.g., `dataset::Dataset`.
+This also matches the Python convention where module names are typically lower-case, thus reducing the differences between the C++ and Python APIs.
+
+#### The Python module name is `dataset`
+
+##### Rationale
+
+This is a corollary from the namespace name.
+
+#### The top level namespace and module is `mantid`
+
+##### Rationale
+
+Currently we have `namespace Mantid` in C++ while the Python module is `mantid`.
+Changing the namespace name would improve consistency.
+Furthermore, this would reduce naming clashes between existing and new code.
+
+##### Questions
+
+Does this make sense?
+Where to we put additional functionality?
+Sibling or children of `dataset`?
+Consider `class Dataset` itself and two hypothetical modules with *additional* functionality that is not part of this library.
+Our options would include:
+1. `mantid::dataset::Dataset`, `mantid::dataset::algorithm`, `mantid::dataset::io`.
```

Comment:
As commented above, I think this would be fine. Extensions like `convert` could go in the `mantid` library and be left out of `dataset`.